### PR TITLE
[fix](scan) Incorrect scan keys lead to wrong query results.

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes

```
mysql [doris_14555]>select * from table_9436528_3;
+------+------+------+------+------------------------+--------------------+------+
| col1 | col2 | col3 | col5 | col4                   | col6               | col7 |
+------+------+------+------+------------------------+--------------------+------+
| -100 |    1 |  -82 |    1 | 2024-02-16 04:37:37.00 | -1299962421.904282 | NULL |
| -100 |    1 |   92 |    1 | 2024-02-16 04:37:37.00 |   23423423.0324234 | NULL |
| -100 |    0 |  -82 |    0 | 2023-11-11 10:49:43.00 |   840968969.872149 | NULL |
```
wrong result:
```
mysql [doris_14555]>select * from table_9436528_3 where col1 <= -100 and col2 in (true, false) and col3 = -82;
+------+------+------+------+------------------------+--------------------+------+
| col1 | col2 | col3 | col5 | col4                   | col6               | col7 |
+------+------+------+------+------------------------+--------------------+------+
| -100 |    1 |  -82 |    1 | 2024-02-16 04:37:37.00 | -1299962421.904282 | NULL |
| -100 |    1 |   92 |    1 | 2024-02-16 04:37:37.00 |   23423423.0324234 | NULL |
+------+------+------+------+------------------------+--------------------+------+
```

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by cambyzju** [comment]
does 2.1 and 2.0 also have this problem?

File: regression-test/suites/correctness/test_scan_keys_with_bool_type.groovy [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    """
+
+    sql """ insert into test_scan_keys_with_bool_type values
+        ( -100 ,    0 ,  -82 ,    0 , '2023-11-11 10:49:43.00' ,   840968969.872149 , NULL ),
+        ( -100 ,    1 ,  -82 ,    1 , '2024-02-16 04:37:37.00' , -1299962421.904282 , NULL ),
+        ( -100 ,    1 ,   92 ,    1 , '2024-02-16 04:37:37.00' ,   23423423.0324234 , NULL );
+    """
+
+    qt_select1 " select * from test_scan_keys_with_bool_type order by 1, 2, 3, 4, 5, 6, 7; "
+    qt_select2 " select * from test_scan_keys_with_bool_type where col1 <= -100 and col2 in (true, false) and col3 = -82 order by 1, 2, 3, 4, 5, 6, 7; "
```
Comment (yiguolei) [inline regression-test/suites/correctness/test_scan_keys_with_bool_type.groovy:45]: 1. col2 int,  col2 in (1,2) 
2. col2 char, col2 in ('a','b')

File: regression-test/suites/correctness/test_scan_keys_with_bool_type.groovy [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            `col1` tinyint NOT NULL,
+            `col2` boolean NOT NULL,
+            `col3` tinyint NOT NULL,
+            `col5` boolean REPLACE NOT NULL,
+            `col4` datetime(2) REPLACE NOT NULL,
+            `col6` double REPLACE_IF_NOT_NULL NULL,
+            `col7` datetime(3) REPLACE_IF_NOT_NULL NULL
+        ) ENGINE=OLAP
+        AGGREGATE KEY(`col1`, `col2`, `col3`)
+        DISTRIBUTED BY HASH(`col1`, `col2`, `col3`) BUCKETS 4
```
Comment (yiguolei) [inline regression-test/suites/correctness/test_scan_keys_with_bool_type.groovy:32]: set buckets to 1
disable compaction

**Comment by mrhhsg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by eldenmoon — APPROVED** [review]
LGTM

**Pull request merged**